<a href="https://colab.research.google.com/github/davisbove/GSB_544/blob/main/PA_errors_python.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Overview

In this PA, you will will be creating and manipulating basic objects to uncover a top secret message.

### Hint 1

As you work through this PA, you will encounter some code that does not work as you want it to.  Don't despair!  Errors (when R is unable to run your code) and bugs (when the code doesn't do what you hoped) are a natural part of coding.  Even the best of the best deal with these problems regularly - learning to track down the issue is a skill that you can learn and practice.

Google is your friend!  If an error is confusing, copy it into Google and see what other people are saying.  If you don't know how to do something, search for it.

### Hint 2

Just because there is no error message doesn't mean everything went smoothly.  You should be regularly printing out your objects, to see they contain the information and format that you expect.

For example, if you have an object named `my_thing`, you might first want to simply look at the object by typing its name.

However, you might also want to make sure it is the right **object type** and/or the right **object structure**.


In [ ]:
type(my_thing)
len(my_thing)

## Part One: Setup

Each of the following code chunks will cause an error and/or do the desired task incorrectly.  Find the mistake, and correct it to complete the intended action.




1. Create lists containing the upper case letters, lower case letters, and some punctuation marks.


In [1]:
lower_case = ["a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m", "n", "o", "p", "q", "r", "s", "t", "u", "v", "w", "x", "y", "z"]

upper_case = ["A", "B", "C", "D", "E", "F", "G", "H" "I", "J", "K", "L", "M", "N", "O", "P", "Q", "R", "S", "T", "U", "V", "W", "X", "Y", "Z"]

punctuation = [".", ",", "!", "?", "'", "\"", "(", ")", " ", "-", ";", ":"]

2. Make one long list containing all the symbols.  Find the length of this list.


In [3]:
all_symbols = lower_case + upper_case + punctuation

len(all_symbols)

63

3. Repeat the previous steps, but instead of using *list* type objects, use **numpy arrays**.  Don't forget to **import** the numpy library!


In [22]:
help(np.concatenate)

Help on function concatenate in module numpy:

concatenate(...)
    concatenate((a1, a2, ...), axis=0, out=None, dtype=None, casting="same_kind")
    
    Join a sequence of arrays along an existing axis.
    
    Parameters
    ----------
    a1, a2, ... : sequence of array_like
        The arrays must have the same shape, except in the dimension
        corresponding to `axis` (the first, by default).
    axis : int, optional
        The axis along which the arrays will be joined.  If axis is None,
        arrays are flattened before use.  Default is 0.
    out : ndarray, optional
        If provided, the destination to place the result. The shape must be
        correct, matching that of what concatenate would have returned if no
        out argument were specified.
    dtype : str or dtype
        If provided, the destination array will have this dtype. Cannot be
        provided together with `out`.
    
        .. versionadded:: 1.20.0
    
    casting : {'no', 'equiv', 'safe', '

In [21]:
import numpy as np
from numpy import concatenate

lower_case = np.array(["a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m", "n", "o", "p", "q", "r", "s", "t", "u", "v", "w", "x", "y", "z"])

upper_case = np.array(["A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L", "M", "N", "O", "P", "Q", "R", "S", "T", "U", "V", "W", "X", "Y", "Z"])

punctuation = np.array([".", ",", "!", "?", "'", "\"", "(", ")", " ", "-", ";", ":"])

## Use np.concatenate to combine into long array
all_symbols = np.concatenate((lower_case, upper_case, punctuation), axis = 0)

4. Make a **pandas data frame**, with one column called "Symbol" containing the `all_symbols` array.


In [6]:
import pandas as pd

dat = pd.read_csv("https://gist.githubusercontent.com/slopp/ce3b90b9168f2f921784de84fa445651/raw/4ecf3041f0ed4913e7c230758733948bc561f434/penguins.csv")

dat

#dat.info()

rowid    species     island  bill_length_mm  bill_depth_mm  \
0        1     Adelie  Torgersen            39.1           18.7   
1        2     Adelie  Torgersen            39.5           17.4   
2        3     Adelie  Torgersen            40.3           18.0   
3        4     Adelie  Torgersen             NaN            NaN   
4        5     Adelie  Torgersen            36.7           19.3   
..     ...        ...        ...             ...            ...   
339    340  Chinstrap      Dream            55.8           19.8   
340    341  Chinstrap      Dream            43.5           18.1   
341    342  Chinstrap      Dream            49.6           18.2   
342    343  Chinstrap      Dream            50.8           19.0   
343    344  Chinstrap      Dream            50.2           18.7   

     flipper_length_mm  body_mass_g     sex  year  
0                181.0       3750.0    male  2007  
1                186.0       3800.0  female  2007  
2                195.0       3250.0  female  2007  
3                  NaN          NaN     NaN  2007  
4                193.0       3450.0  female  2007  
..                 ...          ...     ...   ...  
339              207.0       4000.0    male  2009  
340              202.0       3400.0  female  2009  
341              193.0       3775.0    male  2009  
342              210.0       4100.0    male  2009  
343              198.0       3775.0  female  2009  

[344 rows x 9 columns]

In [18]:
import pandas as pd



symbols_df = pd.DataFrame(all_symbols)

symbols_df

symbols_df.rename(columns =  {0: "Symbol"})

Symbol
0       a
1       b
2       c
3       d
4       e
..    ...
59      )
60       
61      -
62      ;
63      :

[64 rows x 1 columns]

## Part Two: Decoding the secret message.


In [29]:
top_secret = pd.read_csv("https://www.dropbox.com/s/k72h1zewk4gtqep/PA_Secret_Code?dl=1", header=None)

top_secret
top_secret = top_secret[0].to_numpy()
top_secret

array([0.39104271, 0.38644508, 0.37131377, 0.33881179, 0.37131377,
       0.39695721, 0.35479413, 0.37573224, 0.3693288 , 0.37404822,
       0.37573224, 0.37490299, 0.37490299, 0.40018665, 0.40307943,
       0.37490299, 0.38531114, 0.38024665, 0.40542   , 0.39915133,
       0.38644508, 0.3858839 , 0.40757858, 0.39457479, 0.3924204 ,
       0.39196872, 0.40368332, 0.39330231, 0.40052335, 0.41099418,
       0.40597557, 0.39915133, 0.40214824, 0.39695721, 0.40246212,
       0.41362081, 0.41403541, 0.50672129, 0.46904194, 0.4189279 ,
       0.40513799, 0.50677976, 0.41403541, 0.46026337, 0.44516294,
       0.41006039, 0.50685267, 0.40485308, 0.43565499, 0.40982195,
       0.50691085, 0.41751216, 0.42308635, 0.43400049, 0.40933885,
       0.43181048, 0.4148465 , 0.50701239, 0.42481402, 0.43765951,
       0.50705579, 0.48348599, 0.42352845, 0.41751216, 0.50711356,
       0.45815983, 0.41403541, 0.41504562, 0.41927084, 0.50976041,
       0.50719998, 0.4658871 , 0.43527447, 0.43546538, 0.50725

### Hint

Recall that basic math operations are **vectorized**.  This means you don't have to repeat the process for each item in a list; you can apply the math to the whole thing at once.

### Instructions

By altering this top secret set of numbers, you will be able to create a message.  Write your own code to complete the steps below.

1. Add 14 to every number.
2. Multiply every number by 18, then subtract 257.
3. Exponentiate every number.  (That is, do e^[number].  You may have to Google how to do this!)
4. Square every number.

**Checkpoint:** Headquarters has informed you that at this stage of decoding, there should be 352 numbers in the secret message that are below 17.  Write code to check if this is true for your data.


5. Turn your one-dimensional array of numbers into a two-dimensional array (matrix) with 5 columns.
6. Separately from your top secret numbers, create an array of all the even numbers between 1 and 382. Name it "evens". That is, "evens" should contain 2, 4, 6, 8 ..., 382.
7. Subtract the "evens" vector from the first column of your secret message matrix.
8. Subtract 100 from all numbers in the 18-24th rows of the 3rd column.
9. Multiply all numbers in the 4th and 5th column by 2.
10. Turn your matrix back into a one dimensional array.


**Checkpoint:** Headquarters has informed you that at this stage of decoding, all numbers in indices 500 and beyond are below 100.  Write code to check if your numbers fulfill this requirement.

11. Take the square root of all numbers in indices 37 to 464.
12. Round all numbers to the nearest whole number.
13. Replace all instances of the number 39 with 20.

**Checkpoint:** Headquarters has informed you that your final message should have 344 even numbers.  Write code to check if your numbers fulfill this requirement.

## Part 3: The secret message!

Run the following to discover the final message!  Google the first line of this message, if you do not recognize it, to see what it is.


In [145]:

top_secret
top_secret_2 = top_secret +14
top_secret_3 = ((top_secret_2) *18 ) -257
top_secret_4 = np.exp(top_secret_3)
top_secret_5 = (top_secret_4) **2
top_secret_17 = top_secret_5 < 17
np.count_nonzero(top_secret_17 == True)

352

In [159]:




top_secret_mat = np.reshape(top_secret_5, (191, 5))


evens = np.arange(2, 384, 2)

top_secret_mat[:, 0] = top_secret_mat[:, 0] - evens
top_secret_mat
top_secret_mat[18:24, 2 ] -= 100
top_secret_mat[:, 3] *= 2
top_secret_mat[:, 4] *= 2
top_secret_mat[:, 3]
top_secret_final = np.reshape(top_secret_mat, ( 955 ,1))
top_secret_final[500:955]<100
top_secret_final[37:464]**= .5
np.round(top_secret_final)
top_secret_final.astype(int)
thirty_nines = top_secret_final == 39
top_secret_final.loc(thirty_nines)


AttributeError: ignored

In [ ]:
''.join(map(str, symbols_df.loc[top_secret]['Symbol']))